### PR TITLE
feat(settings): Add device selection for microphone/speaker

### LINKED ISSUE #1174: feat(calling): allow user to change their speaker/microphone
if for some reason the default speaker/microphone doesn't work for someone, they will need to change it. Blink already provides an API for this. 

image

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Adds ability to select microphones/speaker devices from audio settings

### Which issue(s) this PR fixes 🔨

- Resolve #1174

### CODE REVIEW

File: common/src/warp_runner/manager/commands/blink_commands.rs [hunk 1]
```diff
@@ -36,6 +41,20 @@ pub enum BlinkCmd {
     UnmuteSelf {
         rsp: oneshot::Sender<Result<(), warp::error::Error>>,
     },
+    #[display(fmt = "GetAllMicrophones")]
+    GetAllMicrophones { rsp: oneshot::Sender<DeviceResult> },
+    #[display(fmt = "GetAllMicrophones")]
```
Comment (lgmarchi) [inline common/src/warp_runner/manager/commands/blink_commands.rs]: Display Here should be `SetMicrophone`

File: common/src/warp_runner/manager/commands/blink_commands.rs [hunk 2]
```diff
@@ -36,6 +41,20 @@ pub enum BlinkCmd {
...
+    SetMicrophone {
+        device_name: String,
+        rsp: oneshot::Sender<Result<(), warp::error::Error>>,
+    },
+    #[display(fmt = "GetAllMicrophones")]
```
Comment (lgmarchi) [inline common/src/warp_runner/manager/commands/blink_commands.rs]: Display Here should be `GetAllSpeakers`

File: common/src/warp_runner/manager/commands/blink_commands.rs [hunk 3]
```diff
@@ -36,6 +41,20 @@ pub enum BlinkCmd {
...
+    GetAllSpeakers { rsp: oneshot::Sender<DeviceResult> },
+    #[display(fmt = "GetAllMicrophones")]
```
Comment (lgmarchi) [inline common/src/warp_runner/manager/commands/blink_commands.rs]: Display Here should be `SetSpeaker`

File: ui/src/components/settings/sub_pages/audio.rs [hunk 4]
```diff
@@ -2,21 +2,129 @@ use std::vec;
...
+                            .send(WarpCmd::Blink(BlinkCmd::GetAllMicrophones { rsp: tx }))
+                        {
+                            log::error!("failed to send blink command");
+                            continue;
+                        }
+
+                        match rx.await {
+                            Ok(res) => {
+                                state.write_silent().settings.input_device = res.1;
+                                *input_devices.write() = res.0.unwrap();
```
Comment (lgmarchi) [inline ui/src/components/settings/sub_pages/audio.rs]: Should not be better to use `unwrap_or_default()` here?

File: ui/src/components/settings/sub_pages/audio.rs [hunk 5]
```diff
@@ -2,21 +2,129 @@ use std::vec;
...
+                            log::error!("failed to send blink command");
+                            continue;
+                        }
+
+                        match rx.await {
+                            Ok(_) => {
+                                state.write_silent().settings.input_device = Some(device);
+                            }
+                            Err(e) => {
+                                log::error!("warp_runner failed to set output device: {e}");
```
Comment (lgmarchi) [inline ui/src/components/settings/sub_pages/audio.rs]: Error here should be input device in the log

**Review by lgmarchi — CHANGES_REQUESTED** [review]
LGTM. Just fix some texts in the PR and 2 unwrap() code.

File: common/src/warp_runner/manager/commands/blink_commands.rs [hunk 6]
```diff
@@ -5,6 +5,11 @@ use warp::{blink::AudioCodec, crypto::DID};
 
 use crate::warp_runner::Calling;
 
+type DeviceResult = (
```
Comment (sdwoodbury) [inline common/src/warp_runner/manager/commands/blink_commands.rs]: I recommend changing this to a struct with fields `available_devices` and `selected_device`. Eliminating the tuple would result in one less thing to memorize.

File: ui/src/components/settings/sub_pages/audio.rs [hunk 7]
```diff
@@ -27,9 +135,11 @@ pub fn AudioSettings(cx: Scope) -> Element {
                 section_description: get_local_text("settings-audio.input-device-description"),
                 no_border: true,
                 Select {
-                    initial_value: "Default".into(),
-                    options: vec!["Default".into()],
-                    onselect: move |_| {}
+                    initial_value: state.read().settings.input_device.as_ref().cloned().unwrap_or("Default".into()),
```
Comment (sdwoodbury) [inline ui/src/components/settings/sub_pages/audio.rs]: `Default` should probably be `default`. I hope it doesn't vary by platform. I have some changes planned for Blink to use a separate code path when the device name is "default"/"Default"

**Review by sdwoodbury — APPROVED** [review]
this looks fine. Note that there's a branch in Warp called `feat/audio-test` which will allow the user to test the speaker/microphone setup before finalizing the audio configuration. The Blink API will change slightly. 

I suspect that eventually, we won't want to apply the configuration changes until the user navigates away from this page but that will depend on whatever UI Kevin designs.

**Review by lgmarchi — APPROVED** [review]
LGTM.

**Review by stavares843 — CHANGES_REQUESTED** [review]
very neat 🎉 

something i noticed is

i have a macbook pro
then i connected by exterior headphones and the beats phones as well

when i deactivate them and no longer are connected to my macbook pro they still appear listed in uplink

Captura de ecrã 2023-09-13, às 17 00 59
Captura de ecrã 2023-09-13, às 17 01 06

**Comment by Flemmli97** [comment]
> very neat 🎉
> 
> something i noticed is
> 
> i have a macbook pro then i connected by exterior headphones and the beats phones as well
> 
> when i deactivate them and no longer are connected to my macbook pro they still appear listed in uplink
> 
> Captura de ecrã 2023-09-13, às 17 00 59 Captura de ecrã 2023-09-13, às 17 01 06

did you refresh the page? aka go to another page and back? or did you do it while on the settings page?

**Comment by stavares843** [comment]
> did you do it while on the settings page?

while in the settings page yes

**Comment by sdwoodbury** [comment]
`cpal` doesn't provide callbacks when a device is disconnected. Blink only emits an event if there's an ongoing call and the audio device is disconnected because that results in an error. I think the best we can do is call blink's `get_available_microphones/speakers` functions every few seconds while this page is open to update the list.

**Review by stavares843 — APPROVED** [review]

**Pull request merged**